### PR TITLE
src/crimson/osd/scrub: fix the null pointer error

### PR DESCRIPTION
**Pull request by guojidan**
Fixes: https://tracker.ceph.com/issues/66405

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by guojidan** [comment]
we can assert that `maybe_obj->negative` is false directly because "impossible since chunky scrub was introduced".

File: src/crimson/osd/scrub/scrub_validator.cc [hunk 1]
```diff
@@ -60,9 +60,10 @@ shard_evaluation_t evaluate_object_shard(
   if (from == policy.primary) {
     ret.shard_info.primary = true;
   }
-  if (!maybe_obj || maybe_obj->negative) {
-    // impossible since chunky scrub was introduced
-    ceph_assert(!maybe_obj->negative);
+
+  // impossible since chunky scrub was introduced
+  ceph_assert(!maybe_obj->negative);
```
Comment (Matan-B) [inline src/crimson/osd/scrub/scrub_validator.cc:65]: maybe_obj may be nullptr

**Review by Matan-B — CHANGES_REQUESTED** [review]
The issue is that we check if `maybe_obj` is null and then try to access it in l.65.

```
  if (!maybe_obj) {
    ret.shard_info.set_missing();
    return ret;
  }
  // impossible since chunky scrub was introduced
  ceph_assert(maybe_obj->negative == false);
```

**Comment by athanatos** [comment]
Probably just change the assert to ceph_assert(!maybe_obj || !(maybe_obj->negative));

**Comment by guojidan** [comment]
This assertion is only for throwing an exception because `maybe_obj->negative` must be false. So, I don't think it's necessary to determine whether `maybe_obj->negative` is true or false.

File: src/crimson/osd/scrub/scrub_validator.cc (hunk 1 above)
Comment (guojidan) [reply src/crimson/osd/scrub/scrub_validator.cc:65]: ah yes, Assert should come after the check for nullptr.

**Comment by guojidan** [comment]
> Probably just change the assert to ceph_assert(!maybe_obj || !(maybe_obj->negative));

Changing the position of the assert will do. This can avoid making the code look complex.

**Comment by guojidan** [comment]
jenkins test make check

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test api

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-july-25/

**Pull request merged**